### PR TITLE
Add type name to ValidationError error message

### PR DESCRIPTION
**Pull request by dmontagu**
## Change Summary

Make the string displayed for ValidationError contain the name of the model for easier debugging.

## Related issue number

#677

## Checklist

Codecov seems to be bugging out -- the "uncovered" lines were setting the value of `compiled` in `main.py` (I've marked the tests+coverage checkbox since I think it should be passing).

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `HISTORY.rst` has been updated
  * if this is the first change since a release, please add a new section
  * include the issue number or this pull request number `#<number>`
  * include your github username `@<whomever>`

### CODE REVIEW

File: pydantic/error_wrappers.py [hunk 1]
```diff
@@ -51,8 +51,9 @@ def dict(self, *, loc_prefix: Optional[Tuple[str, ...]] = None) -> Dict[str, Any
 class ValidationError(ValueError):
     __slots__ = ('raw_errors',)
 
-    def __init__(self, errors: Sequence[ErrorList]) -> None:
+    def __init__(self, errors: Sequence[ErrorList], obj_or_type: Any) -> None:
```
Comment (samuelcolvin) [inline pydantic/error_wrappers.py]:
```suggestion
    def __init__(self, errors: Sequence[ErrorList], model: Type[Any]) -> None:
```

Then you'll need to call `type(...)` in a couple of places, but I think this makes more sense.

File: pydantic/error_wrappers.py [hunk 2]
```diff
@@ -51,8 +51,9 @@ def dict(self, *, loc_prefix: Optional[Tuple[str, ...]] = None) -> Dict[str, Any
...
         self.raw_errors = errors
+        self.obj_type: Type[Any] = obj_or_type if isinstance(obj_or_type, type) else type(obj_or_type)
```
Comment (samuelcolvin) [inline pydantic/error_wrappers.py]: please add this to slots, also I think `model` or `model_cls` would be a better name.

File: pydantic/main.py [hunk 3]
```diff
@@ -687,8 +687,10 @@ def validate_model(  # noqa: C901 (ignore complexity)
                     errors.append(ErrorWrapper(ExtraError(), loc=f, config=config))
 
     if not raise_exc:
-        return values, fields_set, ValidationError(errors) if errors else None
+        model_type = model if isinstance(model, type) else type(model)
+        return values, fields_set, ValidationError(errors, model_type) if errors else None
 
     if errors:
-        raise ValidationError(errors)
+        model_type = model if isinstance(model, type) else type(model)
```
Comment (dmontagu) [inline pydantic/main.py]: @samuelcolvin Any preference for how to eliminate the duplicated definition of model_type? I could move it outside the conditionals but I didn't want to affect performance in the case where an error is not raised.

File: pydantic/main.py [hunk 4]
```diff
@@ -288,7 +288,7 @@ def __setattr__(self, name, value):
         elif self.__config__.validate_assignment:
             value_, error_ = self.fields[name].validate(value, self.dict(exclude={name}), loc=name)
             if error_:
-                raise ValidationError([error_])
+                raise ValidationError([error_], self)
```
Comment (samuelcolvin) [inline pydantic/main.py]:
```suggestion
                raise ValidationError([error_], type(self))
```
?

Comment (samuelcolvin) [reply pydantic/main.py]: and a test would probably be good.

File: pydantic/main.py (hunk 3 above)
Comment (samuelcolvin) [reply pydantic/main.py]: we're also executing `model_type = ...` above even when it's not required.

Maybe something like

```py
err = None
if errors:
    model_type = model if isinstance(model, type) else type(model)
    err = ValidationError(errors, model_type)

if not raise_exc:
    return values, fields_set, err

if err:
    raise err
```
?

**Pull request merged**